linked_list/linked_list.py

```diff
@@ -1,79 +1,158 @@
+        new_node.next = self.head
+        self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) 
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        current = self.head
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        count = 0
+        current = self.head
+        while current:
+            current = current.next
+            count += 1
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        i = 0
+        current = self.head
+        while current:
+            if i == index:
+                return current.value 
+            current = current.next
+            i += 1
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if not self.head:
+            return None
+        current = self.head
+        while current:
+            if not current.next:
+                return current.value
+            current = current.next   
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: ? O(1)
+    # Space Complexity: ? O(1)
     def add_last(self, value):
-        pass
+        if not self.head:
+            self.head = Node(value)
+            return
+
+        current = self.head
+        # get to the last value in the list
+        while current:
+            if not current.next: #if we're at the last node of the list thennnn
+                # add a new node to the end 
+                current.next = Node(value)
+                return 
+            current = current.next
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: ? O(n)
+    # Space Complexity: ? O(1)
     def find_max(self):
-        pass
+        # check is the list is empty
+        if not self.head:
+            return None
+        # create a varible max_value 
+        max_value = 0

```

Comment:
🤔 Consider an edge case where the linked list has negative values. What if the maximum value is -2? 

Try initializing `max_value` to the head's value instead. Whatever the head's value is, if there's something greater than it, you'll find it during your while loop. 

```suggestion
        max_value = self.head

```